lib/modules/asset/AssetService.ts

```diff
@@ -135,6 +135,97 @@ export class AssetService extends BaseService {
     });
   }
 
+  /**
+   * Update or Create an asset metadata
+   */
+  /**
+   * Update or Create an asset metadata
+   */
+  public async upsert(
+    engineId: string,
+    assetId: string,
+    metadata: Metadata,
+    request: KuzzleRequest
+  ): Promise<KDocument<AssetContent>> {
+    return lock(`asset:${engineId}:${assetId}`, async () => {
+      let asset = null;
+      try {
+        asset = await this.get(engineId, assetId, request);
+      } catch (error) {
+        this.app.log.error(`Asset ${assetId} not found`);
+        asset = {
+          _id: assetId,
+          _source: {
+            groups: [],
+            lastMeasuredAt: 0,
+            linkedDevices: [],
+            measures: undefined,
+            metadata: {},
+            model: "",
+            reference: "",
+          },
+        };
+      }
+
+      const upsertedPayload = await this.app.trigger<EventAssetUpdateBefore>(
+        "device-manager:asset:update:before",
+        { asset, metadata }
+      );
+
+      const _source = asset
+        ? { ...asset._source, metadata: upsertedPayload.metadata }
+        : {
+            groups: [],
+            lastMeasuredAt: 0,
+            linkedDevices: [],
+            measures: undefined,
+            metadata: upsertedPayload.metadata,
+            model: "",
+            reference: "",
+          };
+
+      const upsertedAsset = await this.upsertDocument<AssetContent>(
+        request,
+        {
+          _id: assetId,
+          _source,
+        },
+        {
+          collection: InternalCollection.ASSETS,
+          engineId,
+        },
+        { source: true }
+      );
+
+      await this.assetHistoryService.add<AssetHistoryEventMetadata>(engineId, [
+        {
+          asset: upsertedAsset._source,
+          event: {
+            metadata: {
+              names: Object.keys(flattenObject(upsertedPayload.metadata)),
+            },
+            name: "metadata",
+          },
+          id: upsertedAsset._id,
+          timestamp: Date.now(),
+        },
+      ]);
+
+      await this.app.trigger<EventAssetUpdateAfter>(
+        "device-manager:asset:update:after",
```

Comment:
Same as Line 171

Reply:
This is marked as resolved but this is still update event, it should be upsert

Reply:
After the comments, I then attempted to use the upsertDocument method in the BaseService file, but it wasn't appropriate because in the Asset service file, if you use the upsertDocument method, you would need to reconstruct the logic of the existing create and update methods. Therefore, with the aim of maintaining cleanliness and logic in the upsert method within the AssetService file, I chose to use the already established create and update methods